gtsfm/utils/alignment.py

```diff
@@ -0,0 +1,310 @@
+                logger.debug("Sim(3) Rotation `aRb`: rz=%.2f deg., ry=%.2f deg., rx=%.2f deg.", rz, ry, rx)
+                logger.debug("Sim(3) Translation `atb`: [%.2f, %.2f, %.2f]", atb[0], atb[1], atb[2])
+                logger.debug("Sim(3) Scale `asb`: %.2f", float(best_aSb.scale()))
+
+    aRb = best_aSb.rotation().matrix()
+    atb = best_aSb.translation()
+    rz, ry, rx = Rotation.from_matrix(aRb).as_euler("zyx", degrees=True)
+    logger.info("Sim(3) Rotation `aRb`: rz=%.2f deg., ry=%.2f deg., rx=%.2f deg.", rz, ry, rx)
+    logger.info("Sim(3) Translation `atb`: [%.2f, %.2f, %.2f]", atb[0], atb[1], atb[2])
+    logger.info("Sim(3) Scale `asb`: %.2f", float(best_aSb.scale()))
+
+    best_aTi_: List[Pose3] = [best_aSb.transformFrom(bTi) for bTi in bTi_list]
+    return best_aTi_, best_aSb
+
+
+def align_poses_sim3_robust(aTi_list: List[Pose3], bTi_list: List[Pose3]) -> Tuple[List[Pose3], Similarity3]:
+    """Align two pose graphs via similarity transformation by trying out some samples of pairs of cameras for candidate
+    transforms.
+    Note: poses cannot be missing/invalid.
+
+    We force Sim(3) alignment rather than SE(3) alignment.
+    We assume the two trajectories are of the exact same length.
+
+    Args:
+        aTi_list: reference poses in frame "a" which are the targets for alignment
+        bTi_list: input poses which need to be aligned to frame "a"
+
+    Returns:
+        aTi_list_: transformed input poses previously "bTi_list" but now which
+            have the same origin and scale as reference (now living in "a" frame)
+        aSb: Similarity(3) object that aligns the two pose graphs.
+    """
+    assert len(aTi_list) == len(bTi_list)
+    n_to_align = len(aTi_list)
+    if n_to_align < 2:
+        logger.error("SIM(3) alignment uses at least 2 frames; Skipping")
+        return bTi_list, Similarity3(Rot3(), np.zeros((3,)), 1.0)
+
+    max_possible_hypothesis: int = (n_to_align * (n_to_align - 1)) // 2
+    if max_possible_hypothesis <= MAX_NUM_HYPOTHESIS_FOR_ROBUST_ALIGNMENT:
+        return align_poses_sim3_exhaustive(aTi_list=aTi_list, bTi_list=bTi_list)
+
+    best_pose_auc_5deg = 0
+    best_aSb = Similarity3()
+
+    sample_pose_pairs = np.random.choice(
+        len(aTi_list),
+        size=MAX_NUM_HYPOTHESIS_FOR_ROBUST_ALIGNMENT * 2,
+        replace=True,
+    ).reshape(-1, 2)
+
+    for i, j in sample_pose_pairs:
+        aTi_sample = copy.deepcopy([aTi_list[i], aTi_list[j]])
+        bTi_sample = copy.deepcopy([bTi_list[i], bTi_list[j]])
+
+        _, aSb_candidate = align_poses_sim3(aTi_sample, bTi_sample)
+
+        aTi_candidate_: List[Pose3] = [aSb_candidate.transformFrom(bTi) for bTi in bTi_list]
+
+        candidate_metrics = metric_utils.compute_ba_pose_metrics(
+            gt_wTi_list=aTi_list,
+            computed_wTi_list=aTi_candidate_,
+        )
+        pose_auc_5deg = candidate_metrics.metrics[6].data.item()
```

Comment:
can we make a helper to return the pose AUC from the class, instead of hard coding index 6 which is a bit brittle IMHO

Reply:
Done

Reply:
Marking as resolved, thanks